src/cpp/src/parsers.cpp

```diff
@@ -0,0 +1,299 @@
+            size_t num_chars_to_keep = 0; 
+            // We must be sure that no chunks with the closing tag are included to reason_str.
+            for (size_t i = txt_chunk.size(); i >= 1; --i) {
+                // Get the substring of the i last characters of txt_chunk
+                auto suffix = txt_chunk.substr(txt_chunk.size() - i, i);
+                // If this suffix is a prefix of m_close_tag, we need to keep it in the cache.
+                if (m_close_tag.find(suffix) == 0) {
+                    num_chars_to_keep = i;
+                    break;
+                }
+            }
+
+            // If the suffix is a prefix of m_close_tag, we store it in the cache to detect if </think> is split between several delta_text pieces.
+            if (num_chars_to_keep > 0) {
+                m_text_cache = txt_chunk.substr(txt_chunk.size() - num_chars_to_keep, num_chars_to_keep);
+                reason_str += txt_chunk.substr(0, txt_chunk.size() - num_chars_to_keep);
+            } else {
+                reason_str += txt_chunk;
+                m_text_cache = "";
+            }
+
+            if (!m_keep_original_content) {
+                delta_text = "";
+            }
+            msg["reasoning_content"] = reason_str;
+        } else {
+            // Think tag was not opened yet and not found in the current delta_text.
+            // Accumulate text in the cache to detect if <think> is split between several delta_text pieces.
+            m_text_cache += delta_text;
+        }
+        
+        return delta_text;
+    }
+};
+
+ReasoningParser::ReasoningParser(bool starts_with_thinking, bool keep_original_content) {
+    m_impl = std::make_shared<ReasoningParserImpl>(starts_with_thinking, keep_original_content);
+}
+
+std::string ReasoningParser::parse(
+    JsonContainer& msg,
+    const std::string& previous_text, 
+    std::string& delta_text,
+    const std::optional<std::vector<int64_t>>& previous_tokens, 
+    const std::optional<std::vector<int64_t>>& delta_tokens
+) {
+    return m_impl->parse(msg, previous_text, delta_text, previous_tokens, delta_tokens);
+}
+
+JsonContainer Llama32PythonicToolParser::parse(JsonContainer& input) {
+    // Input example
+    // string input = "[get_weather(location='New York, NY', unit='celsius')]<|eom_id|>";
+
+    // Regex to capture the [...] part
+    std::smatch m;
+    const std::string& text = input["content"].get_string();
+    std::regex r(R"(\[.*?\])");
+    if (!std::regex_search(text, m, r)) {
+        return input;
+    }
+
+    // Strip outer [ ]
+    std::string call = m.str().substr(1, m.str().size() - 2);
+
+    // Split function name and arguments
+    input["tool_calls"] = JsonContainer::array();
+    
+    size_t pos = call.find('(');
+    std::string name = call.substr(0, pos);
+    std::string args = call.substr(pos + 1, call.size() - pos - 2); // inside (...)
+    
+    
+    JsonContainer kv;
+    // Parse arguments of the form key='value'
```

Comment:
```suggestion
    // Parse arguments of the form key="value"
```
I think it won't process `'` symbol.

Reply:
fixed